src/StorageRegistry.sol

```diff
@@ -0,0 +1,321 @@
+        /* Ensure that the L2 sequencer is up. */
+        (, int256 sequencerUp, uint256 startedAt,,) = uptimeFeed.latestRoundData();
+        if (sequencerUp != 0) revert SequencerDown();
+
+        /* If the L2 sequencer recently restarted, ensure the grace period has elapsed. */
+        uint256 timeSinceUp = block.timestamp - startedAt;
+        if (timeSinceUp < L2_DOWNTIME_GRACE_PERIOD) revert GracePeriodNotOver();
+
+        /* Get and validate the Chainlink ETH/USD price. */
+        (uint80 roundId, int256 answer,, uint256 updatedAt, uint80 answeredInRound) = priceFeed.latestRoundData();
+        if (answer <= 0) revert InvalidPrice();
+        if (updatedAt == 0) revert IncompleteRound();
+        if (answeredInRound < roundId) revert StalePrice();
+
+        return uint256(answer);
+    }
+
+    function _price(uint256 units, uint256 usdPerUnit, uint256 ethPerUsd) internal pure returns (uint256) {
+        return units * usdPerUnit * 1e18 / ethPerUsd;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              OWNER ACTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Credit a list of fids with free storage units. Only callable by owner.
+     *
+     * @param fids  An array of fids.
+     * @param units Number of storage units per fid.
+     */
+    function batchCredit(uint256[] calldata fids, uint256 units) external onlyOwner whenNotDeprecated {
+        uint256 totalUnits = fids.length * units;
+        if (rentedUnits + totalUnits > maxUnits) revert ExceedsCapacity();
+        rentedUnits += totalUnits;
+        for (uint256 i; i < fids.length; ++i) {
+            emit Rent(msg.sender, fids[i], units);
+        }
+    }
+
+    /**
+     * @notice Change the USD price per storage unit.
+     *
+     * @param usdPrice The new unit price in USD. Fixed point value with 8 decimals.
+     */
+    function setPrice(uint256 usdPrice) external onlyOwner {
+        emit SetPrice(usdUnitPrice, usdPrice);
+        usdUnitPrice = usdPrice;
+    }
+
+    /**
+     * @notice Change the maximum supply of storage units.
+     *
+     * @param max The new maximum supply of storage units.
+     */
+    function setMaxUnits(uint256 max) external onlyOwner {
+        emit SetMaxUnits(maxUnits, max);
+        maxUnits = max;
+    }
+
+    /**
+     * @notice Change the deprecationTimestamp.
+     *
+     * @param timestamp The new deprecationTimestamp.
+     */
+    function setDeprecationTimestamp(uint256 timestamp) external onlyOwner {
+        emit SetDeprecationTimestamp(deprecationTimestamp, timestamp);
+        deprecationTimestamp = timestamp;
+    }
+
+    /**
+     * @notice Withdraw a specified amount of ether from the contract balance to a given address.
+     *
+     * @param to     Address of recipient.
+     * @param amount The amount of ether to withdraw.
+     */
+    function withdraw(address to, uint256 amount) external onlyOwner {
```

Comment:
security: thoughts on having two levels of admins? one that can move funds (for operations) and one that can change where the funds go (vault pattern)